### PR TITLE
Implement DSL for LogPoller

### PR DESCRIPTION
**Pull request by EasterTheBunny**
This commit implements the query DSL within LogPoller by remapping comparer filters to
topic and word filters. A parser then creates an SQL statement from the remapped events
and queries the log poller database directly.

### CODE REVIEW

File: core/chains/evm/logpoller/log_poller.go [hunk 1]
```diff
@@ -1344,6 +1345,6 @@ func EvmWord(i uint64) common.Hash {
 	return common.BytesToHash(b)
 }
 
-func (lp *logPoller) FilteredLogs(queryFilter query.KeyFilter, sortAndLimit query.LimitAndSort) ([]Log, error) {
-	return lp.orm.FilteredLogs(queryFilter, sortAndLimit)
+func (lp *logPoller) FilteredLogs(ctx context.Context, queryFilter query.KeyFilter, sortAndLimit query.LimitAndSort, mapper *EventFilterMapper) ([]Log, error) {
```
Comment (ilija42) [inline core/chains/evm/logpoller/log_poller.go:1348]: Why do we need `EventFilterMapper`, all of this information should be in the filters. I wrote a function that maps chain agnostic filters to chain specific ones

Comment (EasterTheBunny) [reply core/chains/evm/logpoller/log_poller.go:1348]: I did this because I needed to consolidate the query remapping with the ORM visitor in the same package to avoid circular dependencies.

File: core/chains/evm/logpoller/expression.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+	sort := make([]string, len(q))
+
+	for idx, sorted := range q {
+		var name string
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
```
Comment (ilija42) [inline core/chains/evm/logpoller/expression.go:90]: This should translate to whatever uniquely identifies the order of things on that chain. Something like block-txID-logID... whatever the EVM equivalent is

Comment (ilija42) [reply core/chains/evm/logpoller/expression.go:90]: Cursor is a separate thing that should be that plus a proper psql cursor? Not completely sure how that looks in practice, but the use case is product can tell the chain reader continue from here

File: core/chains/evm/logpoller/log_poller.go (hunk 1 above)
Comment (ilija42) [reply core/chains/evm/logpoller/log_poller.go]: Thats too bad, I think the remapping should stay out of log poller, must be some other approach to avoid circular dependency here?

File: core/chains/evm/logpoller/expression.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	asGroup := len(expressions) > 1
+	for _, exp := range expressions {
+		if exp.IsPrimitive() {
+			v := &pgDSLParser{
+				chainID: chainID,
+				args:    args,
+			}
+
+			// do as primitive
+			exp.Primitive.Accept(v) // not sure yet what to do with this
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/expression.go:152]: I think this is an old comment. I can remove it.

File: core/chains/evm/logpoller/expression.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		fieldsOut = []string{selectAll}
+	}
+
+	if q.WithCursor {
+		fieldsOut = append(fieldsOut, fmt.Sprintf("%s AS %s", cursorFieldName, cursorFieldAlias))
+	}
+
+	// TODO: if sort by has sequence, add sequence to fields; still don't know composes a sequence but it's different from a cursor
+	if q.WithSequence {
+		fieldsOut = append(fieldsOut, fmt.Sprintf("id AS %s", sequenceFieldAlias))
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/expression.go:34]: Since the sequence/cursor is a constructed field, the field name doesn't exist in the table. It must be constructed in the SELECT fields like `CONCAT(...) AS sequence` where `sequence` is the alias field name. In this case `id` is wrong and doesn't need to be aliased. `id` needs to be replaced with the `CONCAT` bit instead.

File: core/chains/evm/logpoller/expression.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	}
+}
+
+func makeExpression(op query.BoolOperator, expressions []query.Expression, args *namedArgs, chainID *big.Int) string {
+	parts := []string{}
+
+	asGroup := len(expressions) > 1
+	for _, exp := range expressions {
+		if exp.IsPrimitive() {
+			v := &pgDSLParser{
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/expression.go:146]: It would be better on resources by a small margin, but I don't know about cleaner. Worth a discussion though.

File: core/chains/evm/logpoller/expression.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilija42) [reply core/chains/evm/logpoller/expression.go]: It doesn't need to be a field, it just needs to convert to something like ORDER BY (block_number, log_index)`

[the remark this answers is not in the document]
Comment (ilija42) [reply core/chains/evm/logpoller/expression.go]: This is what log poller is doing right now, so I think we can do it here too

File: core/chains/evm/logpoller/parser.go [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+func (p *dslQueryProps) hasCursorLimit() bool {
+	return p.Limiter.Limit.Cursor != nil && p.Limiter.Limit.CursorDirection != nil
+}
+
+func (p *dslQueryProps) hasSequenceSort() bool {
+	for _, order := range p.Limiter.SortBy {
+		switch order.(type) {
+		case query.SortBySequence:
+			return true
+		default:
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:42]: this is undefined behaviour which should probably throw an error

File: core/chains/evm/logpoller/parser.go [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+
+	if props.hasCursorLimit() {
+		var op string
+		switch *props.Limiter.Limit.CursorDirection {
+		case query.Following:
+			op = ">"
+		case query.Previous:
+			op = "<"
+		default:
+			op = "<"
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:246]: should this throw an error?

File: core/chains/evm/logpoller/parser.go [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+		var name string
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			name = sequenceFieldAlias
+		case query.SortByTimestamp:
+			name = timestampFieldName
+		default:
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:284]: not sure if we should tolerate undefined sort queries

File: core/chains/evm/logpoller/parser.go [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+	}
+}
+
+func orderToString(dir query.SortDirection) string {
+	switch dir {
+	case query.Asc:
+		return "ASC"
+	case query.Desc:
+		return "DESC"
+	default:
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:392]: same as other defaults

File: core/chains/evm/logpoller/parser.go [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+	v.expression = fmt.Sprintf(
+		"%s %s :%s",
+		timestampFieldName,
+		cmpOpToString(p.Operator),
+		v.args.Add(timestampFieldName, p.Timestamp),
+	)
+}
+
+func (v *pgDSLParser) TxHashPrimitives(p query.TxHashPrimitive) {
+	v.expression = fmt.Sprintf(
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:91]: Not sure where is the best place to validate thath txHash is common.Hash. Maybe we should have evm specific filter that this gets remapped to? Same as ValuComparer?

File: core/chains/evm/logpoller/parser.go [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+		"%s = :%s",
+		txHashFieldName,
+		v.args.Add(txHashFieldName, p.TxHash),
+	)
+}
+
+func (v *pgDSLParser) VisitAddressFilter(p *addressFilter) {
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:98]: Same comment as for TxHashPrimitive

File: core/chains/evm/logpoller/parser.go [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+	output := strings.Join(clauses, fmt.Sprintf(" %s ", op.String()))
+
+	if grouped {
+		output = fmt.Sprintf("(%s)", output)
+	}
+
+	return output
+}
+
+type namedArgs struct {
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:332]: Can't we reuse what log poller already has?

```
 type queryArgs struct {
	args map[string]interface{}
	err  []error
}
```

File: core/chains/evm/logpoller/parser.go [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+				cmpOpToString(comp.Operator),
+				placement,
+			)
+		}
+
+		v.expression = strings.Join(comps, " AND ")
+	}
+}
+
+func (v *pgDSLParser) buildQuery(chainID *big.Int, filter query.KeyFilter, limiter query.LimitAndSort) (string, map[string]any, error) {
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:189]: I tend to prefer not using named returns. It actually a common linter to not allow them in code (though we don't have that linter active). My preference is to communicate with types. If you think the types need more definition, I can create new types for the functions to return and make it more clear.

File: core/chains/evm/logpoller/parser.go [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+	// reset transient properties
+	v.args = newNamedArgs()
+	v.expression = ""
+	v.chainID = chainID
+
+	props := &dslQueryProps{
```
[the remark this answers is not in the document]
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:195]: That simplifies it some. Done

File: core/chains/evm/logpoller/parser.go (hunk 12 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:366]: I didn't really want to modify the existing query args, but I'll include a commit just for this and we can see how it goes.

File: core/chains/evm/logpoller/parser.go [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+	}
+
+	return strings.Join(clauses, " "), v.args, nil
+}
+
+func (v *pgDSLParser) selectClause(limiter query.LimitAndSort) string {
+	fields := []string{fmt.Sprintf("%s.*", evmLogsTableName)}
+
+	// if order by has sequence or if cursor limit is used, add sequence to fields
+	if hasCursorLimit(limiter.Limit) || hasSequenceSort(limiter.SortBy) {
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:228]: Limiter has `hasCursorLimit` and `hasSequenceSort` as its methods now

File: core/chains/evm/logpoller/parser.go [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+		blockFieldName,
+		cmp,
+		v.args.withIndexedField(blockFieldName, p.Block),
+	)
+}
+
+func (v *pgDSLParser) Confirmations(p primitives.Confirmations) {
+	switch p.ConfirmationLevel {
+	case primitives.Finalized:
+		v.expression = v.nestedConfQuery(true, 0)
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:66]: shouldn't finalized be -1?

File: core/chains/evm/logpoller/parser.go [hunk 17]
```diff
@@ -0,0 +1,502 @@
...
+	default:
+		confs := evmtypes.Confirmations(p.ConfirmationLevel)
+
+		if p.ConfirmationLevel == primitives.Unconfirmed {
+			confs = evmtypes.Unconfirmed
+		}
+
+		v.expression = v.nestedConfQuery(false, uint64(confs))
+	}
+}
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go:80]: Does this do the same thing?

```
func (v *pgDSLParser) Confirmations(p primitives.Confirmations) {
	switch p.ConfirmationLevel {
	case primitives.Finalized:
		v.expression = v.nestedConfQuery(true, 0)
		v.expression = nestedBlockNumberQuery(evmtypes.Finalized)
	case primitives.Unconfirmed:
		v.expression = nestedBlockNumberQuery(evmtypes.Unconfirmed)
	default:
		v.expression = nestedBlockNumberQuery(evmtypes.Confirmations(p.ConfirmationLevel))
	}
}
```

Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:80]: Yes. Finalized is a special subquery, separate from Unconfirmed and specific confirmation values. Unconfirmed is just an alias for the case of 0 confirmations. That is why the fallthrough exists. I can clean it up a bit though.

File: core/chains/evm/logpoller/parser_test.go [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+)
+
+func assertArgs(t *testing.T, args *queryArgs, numVals int) {
+	values, err := args.toArgs()
+
+	assert.Len(t, values, numVals)
+	assert.NoError(t, err)
+}
+
+func TestDSLParser(t *testing.T) {
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser_test.go:22]: Probably all existing LP queries are worth testing to verify if the SQL generated by the DSL is the same as the original one. (and probably if the results returned from the db are the same)

File: core/chains/evm/logpoller/parser_test.go [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+						BoolOperator: query.OR,
+					}},
+				},
+				BoolOperator: query.AND,
+			}},
+		}
+		limiter := query.LimitAndSort{}
+
+		result, args, err := parser.buildQuery(chainID, expressions, limiter)
+		expected := "SELECT evm.logs.* " +
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser_test.go:277]: It's very hard to understand when parentheses open and close, can we have it formatted in a more user-friendly way in tests? Sth like this for instance (similar to Go code indentation)
```sql
SELECT evm.logs.*
FROM evm.logs
WHERE evm_chain_id = :evm_chain_id
  AND (
        block_timestamp >= :block_timestamp_0
        AND (
                    tx_hash = :tx_hash_0
                    OR block_number <= (SELECT greatest(block_number - :confs_0, 0)
                                        FROM evm.log_poller_blocks
                                        WHERE evm_chain_id = :evm_chain_id
                                        ORDER BY block_number DESC
                                        LIMIT 1)
                )
    )

```

File: core/chains/evm/logpoller/parser.go [hunk 20]
```diff
@@ -0,0 +1,499 @@
...
+	for idx, sorted := range sorting {
+		var name string
+
+		order, err := orderToString(sorted.GetDirection())
+		if err != nil {
+			return "", err
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:307]: I might have already commented on that in one of the previous PRs, but why not have dedicated `SortByBlockAndLogIndex` or maybe some kind of `SortByDefault` or `SortByChainOrder` that will be translated to `block_number, log_index`. This is a default sort for 99% of read queries

File: core/chains/evm/logpoller/parser.go [hunk 21]
```diff
@@ -0,0 +1,499 @@
...
+			if v.err != nil {
+				return
+			}
+		}
+
+		v.expression = strings.Join(comps, " AND ")
+	}
+}
+
+func (v *pgDSLParser) VisitEventTopicsByValueFilter(p *eventByTopicFilter) {
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:173]: It seems that `p.EventSig` is completely ignored here

File: core/chains/evm/logpoller/parser.go [hunk 22]
```diff
@@ -0,0 +1,499 @@
...
+	if len(p.ValueComparers) > 0 {
+		topicIdx := v.args.withIndexedField("topic_index", p.Topic)
+
+		comps := make([]string, len(p.ValueComparers))
+		for idx, comp := range p.ValueComparers {
+			comps[idx], v.err = makeComp(comp, v.args, "topic_value", topicIdx, "topics[:%s] %s :%s")
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:179]: We used to have a validation whether `topicIdx` has values between `[1..3]` or sth like this. I think it should be verified here as well and added to `v.err` field if constraint is violated

Comment (mateusz-sekara) [reply core/chains/evm/logpoller/parser.go:179]: Postgres starts indexing arrays from 1 afaik, so it should be `topicIdx+1`, right? At least this is how it used to work so far

File: core/chains/evm/logpoller/parser.go [hunk 23]
```diff
@@ -0,0 +1,499 @@
...
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			sort[idx] = fmt.Sprintf("block_number %s, tx_hash %s, log_index %s", order, order, order)
+
+			continue
+		case query.SortByTimestamp:
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:313]: It might be a bit yagni, because we never sort by timestamp, right?

File: core/chains/evm/logpoller/parser.go [hunk 24]
```diff
@@ -0,0 +1,499 @@
...
+	}
+
+	var builder strings.Builder
+
+	builder.WriteString(selector)
+	builder.WriteString(from)
+	builder.WriteString(where)
+	builder.WriteString(order)
+
+	return fmt.Sprintf("%s <= %s", blockFieldName, builder.String())
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:154]: I'd drop parentheses from the building blocks and add them here. It would be easier to read and less error-prone when changing nested query
```go
var (
        from     = "FROM evm.log_poller_blocks "
        where    = "WHERE evm_chain_id = :evm_chain_id "
        order    = "ORDER BY block_number DESC LIMIT 1"
        selector string
)

if finalized {
		selector = "SELECT finalized_block_number "
	} else {
		selector = fmt.Sprintf("SELECT greatest(block_number - :%s, 0) ",
			v.args.withIndexedField("confs", confs),
		)
	}

//...

return fmt.Sprintf("%s <= (%s)", blockFieldName, builder.String())
```

File: core/chains/evm/logpoller/parser.go [hunk 25]
```diff
@@ -0,0 +1,499 @@
...
+	ErrUnexpectedCursorFormat = errors.New("unexpected cursor format")
+)
+
+// pgDSLParser is a visitor that builds a postgres query and arguments from a commontypes.QueryFilter
+type pgDSLParser struct {
+	args *queryArgs
+
+	// transient properties expected to be set and reset with every expression
+	expression string
+	err        error
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:40]: Shouldn't we have a slice of errors here? Otherwise, every new error spotted during processing will override existing one, right?

File: core/chains/evm/logpoller/parser.go [hunk 26]
```diff
@@ -0,0 +1,499 @@
...
+
+func (v *pgDSLParser) limitClause(limiter query.LimitAndSort) string {
+	if !limiter.HasCursorLimit() && limiter.Limit.Count == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("LIMIT %d", limiter.Limit.Count)
+}
+
+func (v *pgDSLParser) getLastExpression() (string, error) {
```
Comment (mateusz-sekara) [inline core/chains/evm/logpoller/parser.go:333]: This is a bit odd to me. Get methods should be independent and not perform any cleanup/state reset in the background, right? Are you going to reuse that instance of `pgDSLParser` when building different queries? Probably new struct should be created every time you want to translate DSL to SQL so wiping out this state doesn't give any value, no?

**Review by mateusz-sekara — CHANGES_REQUESTED** [review]
1. How are we going to verify if it doesn't introduce any regression (performance and SQL-wise)?

2. How about implementing all existing LP queries using new structs to test if filter primitives are properly designed? This top-down approach would be very convenient to verify if abstractions match our needs

File: core/chains/evm/logpoller/parser_test.go (hunk 18 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser_test.go:22]: The current querying structure in this PR doesn't intend to satisfy all LP queries. There are some tests on some LP queries to ensure we are on the right track though.

File: core/chains/evm/logpoller/parser_test.go (hunk 19 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser_test.go:277]: Yes, it is, but spaces included in a formatted string like this compared to the actual SQL string are different and cause the tests to fail. The solutions for that are to remove spaces from one or add spaces to the other.

File: core/chains/evm/logpoller/parser.go (hunk 20 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:312]: There is a `SortBySequence` which is intended to be the default for all chains. In the EVM implementation, the sequence is the `block_number` and the `log_index` together.

File: core/chains/evm/logpoller/parser.go (hunk 21 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:171]: This needs to be removed.

File: core/chains/evm/logpoller/parser.go (hunk 22 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:184]: This is structured exactly like the LP function `SelectIndexedLogsTopicGreaterThan`. No modifications are done to the topic index on that one.

File: core/chains/evm/logpoller/parser.go (hunk 23 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:318]: Not in the current LP, but because it exists in the `QueryKey` definition, it is implemented here.

File: core/chains/evm/logpoller/parser.go (hunk 24 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go]: Makes sense.

File: core/chains/evm/logpoller/parser.go (hunk 25 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:38]: The parser builds SQL expressions piece by piece for each `Accept` function call and resets the error and expression values after every call. While each function could produce multiple errors, those could be satisfied the Go way with `errrors.Join`.

File: core/chains/evm/logpoller/parser.go (hunk 26 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:338]: I agree that it is a bit odd, but resetting the expression and error for every call instead of instantiating a new parser made more sense for memory usage. I could rename the function to better describe what is happening if that helps.

File: core/chains/evm/logpoller/parser_test.go (hunk 18 above)
Comment (ilija42) [reply core/chains/evm/logpoller/parser_test.go:22]: @EasterTheBunny Which LP queries are we unable to support? Is it just of confidence levels and SigsExcluded support?

File: core/services/relay/evm/event_binding.go [hunk 27]
```diff
@@ -130,16 +105,27 @@ func (e *eventBinding) GetLatestValue(ctx context.Context, params, into any) err
 }
 
 func (e *eventBinding) QueryKey(ctx context.Context, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]commontypes.Sequence, error) {
-	remappedFilter, err := e.remapFilter(filter)
+	remapped, err := e.remap(filter)
 	if err != nil {
 		return nil, err
 	}
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go:111]: if !e.bound {
		return fmt.Errorf("%w: event not bound", commontypes.ErrInvalidType)
	}

remapped, err := e.remap(filter)
	if err != nil {
		return nil, err
	}

File: core/services/relay/evm/event_binding.go [hunk 28]
```diff
@@ -130,16 +105,27 @@ func (e *eventBinding) GetLatestValue(ctx context.Context, params, into any) err
 }
 
 func (e *eventBinding) QueryKey(ctx context.Context, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]commontypes.Sequence, error) {
-	remappedFilter, err := e.remapFilter(filter)
+	remapped, err := e.remap(filter)
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go:112]:
```suggestion
        if !e.bound {
		return fmt.Errorf("%w: event not bound", commontypes.ErrInvalidType)
	}

	remapped, err := e.remap(filter)
```

Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go:112]: Added

File: core/chains/evm/logpoller/parser_test.go (hunk 18 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser_test.go:22]: // do not include address or event hash; not supported by QueryKey at this time
// these can be tested at the ORM level to verify that the query structure could theoretically be used
SelectBlockByHash
SelectBlockByNumber
SelectLatestBlock
SelectOldestBlock
SelectLogsByBlockRange
GetBlocksRange

// expected support by QueryKey and can include in tests
SelectLogsCreatedAfter (tested)
SelectLogsDataWordBetween (tested)
SelectIndexedLogsCreatedAfter (tested)
SelectLogsWithSigs (tested)
SelectLatestLogByEventSigWithConfs
SelectLogs
SelectLogsDataWordRange
SelectLogsDataWordGreaterThan
SelectIndexedLogsTopicGreaterThan
SelectIndexedLogsTopicRange
SelectIndexedLogs
SelectIndexedLogsByBlockRange
SelectIndexedLogsByTxHash

// have specific pgsql queries and aren't currently supported
SelectLatestLogEventSigsAddrsWithConfs
SelectLatestBlockByEventSigsAddrsWithConfs
SelectIndexedLogsWithSigsExcluding

I can add more tests for the queries that are directly supported by QueryKey to ensure our test coverage is as high as possible.

File: core/chains/evm/logpoller/parser.go (hunk 22 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser.go:184]: I see the issue now. Fixed

File: core/chains/evm/logpoller/parser_test.go (hunk 18 above)
Comment (EasterTheBunny) [reply core/chains/evm/logpoller/parser_test.go:22]: I added new tests to verify all queries supported by `QueryKey`.

Comment (ilija42) [reply core/chains/evm/logpoller/parser_test.go:22]: Ah ok, I completely forgot about non log related functions since they will be supported by separate methods not related to queryKey

**Review by ilija42 — CHANGES_REQUESTED** [review]
There is a failing test in CI
```
--- FAIL: TestDSLParser (0.00s)
    --- FAIL: TestDSLParser/query_for_event_topic (0.00s)
        parser_test.go:247: 
            	Error Trace:	/home/runner/work/chainlink/chainlink/core/chains/evm/logpoller/parser_test.go:247
            	Error:      	Received unexpected error:
            	            	invalid index for topic: 8
            	Test:       	TestDSLParser/query_for_event_topic
FAIL
FAIL	github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller	0.242s
```

File: core/chains/evm/logpoller/parser.go [hunk 29]
```diff
@@ -0,0 +1,504 @@
...
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+const (
+	sequenceFieldAlias = "sequence"
+	sequenceField      = "CONCAT(block_number, '-', tx_hash, '-', log_index)"
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go]: These are unused

File: core/chains/evm/logpoller/parser.go [hunk 30]
```diff
@@ -0,0 +1,504 @@
...
+	blockFieldName     = "block_number"
+	timestampFieldName = "block_timestamp"
+	txHashFieldName    = "tx_hash"
+	eventSigFieldName  = "event_sig"
+	chainIDFieldName   = "evm_chain_id"
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go]: This is unused

File: core/chains/evm/logpoller/parser.go [hunk 31]
```diff
@@ -0,0 +1,504 @@
...
+)
+
+var (
+	ErrUnexpectedCursorFormat = errors.New("unexpected cursor format")
+)
+
+// pgDSLParser is a visitor that builds a postgres query and arguments from a commontypes.QueryFilter
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go]: Maybe expand this comment to make this more clear
`The parser builds SQL expressions piece by piece for each Accept function call and resets the error and expression values after every call.`

File: core/chains/evm/logpoller/parser.go [hunk 32]
```diff
@@ -0,0 +1,504 @@
...
+	args *queryArgs
+
+	// transient properties expected to be set and reset with every expression
+	expression string
+	err        error
+}
+
+var _ primitives.Visitor = (*pgDSLParser)(nil)
+
+func (v *pgDSLParser) Comparator(p primitives.Comparator) {}
```
Comment (ilija42) [inline core/chains/evm/logpoller/parser.go]:
```suggestion
func (v *pgDSLParser) Comparator(_ primitives.Comparator) {}
```

File: core/services/relay/evm/event_binding.go [hunk 33]
```diff
@@ -130,16 +105,31 @@ func (e *eventBinding) GetLatestValue(ctx context.Context, params, into any) err
...
+
+	logs, err := e.lp.FilteredLogs(ctx, remapped, limitAndSort)
 	if err != nil {
 		return nil, err
 	}
 
+	if len(logs) == 0 {
+		return nil, commontypes.ErrNotFound
+	}
+
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go]: I don't think that we should treat this as an error, SelectLogsWithSigs does this:
```
if pkgerrors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
```
Other LogPoller querying methods just return empty slice of logs. Probably makes sense to just return an empty slice of sequences?

File: core/services/relay/evm/event_binding.go [hunk 34]
```diff
@@ -326,6 +316,91 @@ func (e *eventBinding) decodeLog(ctx context.Context, log *logpoller.Log, into a
...
+			Head: commontypes.Head{
+				Identifier: fmt.Sprint(logs[idx].BlockNumber),
+				Hash:       logs[idx].BlockHash.Bytes(),
+				Timestamp:  uint64(logs[idx].BlockTimestamp.Unix()),
+			},
+		}
+
+		var tpVal reflect.Value
+
+		tpInto := reflect.TypeOf(into)
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go]: this would help with readability a bit
```suggestion

		var typeVal reflect.Value

		typeInto := reflect.TypeOf(into)
```

File: core/services/relay/evm/event_binding.go [hunk 35]
```diff
@@ -326,6 +316,91 @@ func (e *eventBinding) decodeLog(ctx context.Context, log *logpoller.Log, into a
...
+		}
+
+		return query.Or(remappedBoolExpressions...), nil
+	}
+
+	return e.remapPrimitive(key, expression)
+}
+
+func (e *eventBinding) remapPrimitive(key string, expression query.Expression) (query.Expression, error) {
+	// remap chain agnostic primitives to chain specific
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go]: maybe move this comment above method definition

**Review by ilija42 — COMMENTED** [review]
minor comments, looks good

File: core/services/relay/evm/event_binding.go (hunk 33 above)
Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go]: Works for me

Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go]: I found out why I originally returned an error here. The interface tests expect a not found error: `assert.True(t, errors.Is(err, types.ErrNotFound))`

Comment (ilija42) [reply core/services/relay/evm/event_binding.go]: Which test, how did log poller pass that test before?

File: core/chains/evm/logpoller/parser.go (hunk 26 above)
Comment (mateusz-sekara) [reply core/chains/evm/logpoller/parser.go:338]: Huh, there is a huge overlap between this PR and https://github.com/smartcontractkit/chainlink/pull/12987, so I commented some items in the latter

File: core/services/relay/evm/event_binding.go (hunk 33 above)
Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go]: `RunQueryKeyInterfaceTests`: the new tests for chain reader

Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go]: Pending PR: https://github.com/smartcontractkit/chainlink-common/pull/508

This PR alters the tests to not return an error, but instead an empty slice.

Comment (ilija42) [reply core/services/relay/evm/event_binding.go]: Approved, then should this also return an empty slice?

Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go]: Correct

**Review by mateusz-sekara — DISMISSED** [review]

File: core/services/relay/evm/event_binding.go [hunk 36]
```diff
@@ -121,7 +121,7 @@ func (e *eventBinding) QueryKey(ctx context.Context, filter query.KeyFilter, lim
 	}
 	remapped.Expressions = append(defaultExpressions, remapped.Expressions...)
 
-	logs, err := e.lp.FilteredLogs(ctx, remapped, limitAndSort)
+	logs, err := e.lp.FilteredLogs(ctx, remapped, limitAndSort, e.contractName+"-"+e.eventName)
```
Comment (ilija42) [inline core/services/relay/evm/event_binding.go:124]: I am not sure if this is enough from observability perspective. We can query for same event from same contract with different filters. We should either pass a queryName from `QueryKey` or somehow turn limitAndSort and filter into a string? Although to make it easier we can do this in a separate PR under a separate ticket.

Comment (EasterTheBunny) [reply core/services/relay/evm/event_binding.go:124]: I think a separate PR would be best to iron out the rest of this detail.

**Review by ilija42 — APPROVED** [review]

**Review by mateusz-sekara — APPROVED** [review]
